src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+module Variational
+
+using ..Core, ..Core.RandomVariables, ..Utilities
+using Distributions, Bijectors
+using ProgressMeter, LinearAlgebra
+using ..Turing: PROGRESS
+using ..Turing: Model, SampleFromPrior, SampleFromUniform
+using ..Turing: Turing
+using Random: AbstractRNG
+
+using ForwardDiff
+using Flux.Tracker, Flux.Optimise
+
+import ..Core: getchunksize, getADtype
+
+export
+    vi,
+    ADVI,
+    ELBO
+
+
+abstract type VariationalInference{AD} end
+
+getchunksize(::T) where {T <: VariationalInference} = getchunksize(T)
+getchunksize(::Type{<:VariationalInference{AD}}) where AD = getchunksize(AD)
+getADtype(alg::VariationalInference) = getADtype(typeof(alg))
+getADtype(::Type{<: VariationalInference{AD}}) where {AD} = AD
+
+abstract type VariationalObjective end
+
+const VariationalPosterior = Distribution{Multivariate, Continuous}
+
+"""
+    rand(vi::VariationalInference, num_samples)
+
+Produces `num_samples` samples for the given VI method using number of samples equal to `num_samples`.
+"""
+function rand(vi::VariationalPosterior, num_samples) end
+
+"""
+    objective(vi::VariationalInference, q::VariationalPosterior, model::Model, args...)
+
+Computes the variational objective to be optimized for a given VI method.
+"""
+function objective(vi::VariationalInference, q::VariationalPosterior, model::Model, num_samples) end
+
+# (::VariationalObjective)(vi::VariationalInference, model::Model, num_samples, args...; kwargs...) = begin
+# end
+
+
+"""
+    grad!(vo, vi::VariationalInference, q::VariationalPosterior, model::Model, θ, out, args...)
+
+Computes the gradients used in `optimize!`. Default implementation is provided for `VariationalInference{AD}` where `AD` is either `ForwardDiffAD` or `TrackerAD`. This implicitly also gives a default implementation of `optimize!`.
```

Comment:
Please reformat for 92 char limit. See style guide :)

Reply:
How do you do this when the signature of the method is spread > 92 chars? E.g.
```
    grad!(vo, vi::VariationalInference, q::VariationalPosterior, model::Model, θ, out, args...)
```
is > 92 chars.

Reply:
Yeah, I would also spread this over multiple lines -- in general prefer to break up stuff over multiple lines than have a single long line, or have a single short line that loses information (i.e. please don't sacrifice information for the sake of brevity)

Definitely don't exceed the 92 char limit, but do feel free to write a docstring that runs over multiple lines. e.g.
```
"""
    foo(args...)

Some descriptive text that ...
runs over 92 characters.

# Arguments:
- 
"""
```
is totally fine.

Reply:
Alright, I'll do that then.

A general question: I believe I've addressed the style issue (other than this one). Should I just resolve your comments as I see fit, or is the idea that I say "I've fixed it" and then you go through and resolve?